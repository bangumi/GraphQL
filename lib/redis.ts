import IORedis from 'ioredis';

import { redisOption } from './config';

export default new IORedis(redisOption);
export const Subscriber = new IORedis(redisOption);
