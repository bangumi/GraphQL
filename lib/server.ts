import * as fs from 'node:fs';
import * as path from 'node:path';

import type { FastifyInstance, FastifyRequest, FastifyServerOptions } from 'fastify';
import { fastify } from 'fastify';
import mercurius from 'mercurius';
import AltairFastify from 'altair-fastify-plugin';
import swagger from '@fastify/swagger';
import type { OpenAPIV3 } from 'openapi-types';

import { schema } from './graphql/schema';
import type { Context } from './graphql/context';
import prisma from './prisma';
import * as auth from './auth';
import * as rest from './rest';
import { pkg, projectRoot } from './config';
import { Security } from './openapi';

export async function createServer(opts: FastifyServerOptions = {}): Promise<FastifyInstance> {
  const server = fastify(opts);

  await server.register(mercurius, {
    schema,
    path: '/v0/graphql',
    graphiql: false,
    allowBatchedQueries: true,
    context: async (request: FastifyRequest): Promise<Context> => {
      return {
        prisma,
        auth: await auth.byHeader(request.headers.authorization),
      };
    },
  });

  await server.register(AltairFastify, {
    path: '/v0/altair/',
    baseURL: '/v0/altair/',
    endpointURL: '/v0/graphql',
    initialSettings: {
      theme: 'dark',
      'plugin.list': ['altair-graphql-plugin-graphql-explorer'],
    },
  });

  const swaggerUI = fs.readFileSync(path.join(projectRoot, './lib/swagger.html'));

  server.get('/v0.5/', (_, res) => {
    void res.type('text/html').send(swaggerUI);
  });

  server.get('/v0.5/openapi.json', () => {
    return server.swagger();
  });

  const openapi: Partial<OpenAPIV3.Document> = {
    info: {
      version: pkg.version,
      title: 'hello',
    },
    components: {
      securitySchemes: {
        [Security.OptionalHTTPBearer]: {
          type: 'http',
          description:
            '不强制要求用户认证，但是可能看不到某些敏感内容内容（如 NSFW 或者仅用户自己可见的收藏）',
          scheme: 'Bearer',
        },
        [Security.HTTPBearer]: {
          type: 'http',
          description: '需要使用 access token 进行认证',
          scheme: 'Bearer',
        },
      },
    },
  };

  await server.register(swagger, {
    openapi,
  });

  await server.register(rest.setup, { prefix: '/v0.5' });

  return server;
}
