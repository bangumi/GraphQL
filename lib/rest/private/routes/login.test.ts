import { afterEach, beforeEach, describe, expect, test } from 'vitest';

import prisma from '../../../prisma';
import redis from '../../../redis';
import { createServer } from '../../../server';

describe('login', () => {
  beforeEach(async () => {
    await redis.flushdb('SYNC');
    await prisma.chii_os_web_sessions.deleteMany();
  });

  afterEach(async () => {
    await redis.flushdb('SYNC');
    await prisma.chii_os_web_sessions.deleteMany();
  });

  test('should failed on too many requests', async () => {
    const app = await createServer();

    const opt = {
      method: 'post',
      url: '/p1/login',
      payload: { email: 'ee', password: 'eepp', 'h-captcha-response': 'fake-response' },
    } as const;

    const login = () => app.inject(opt);

    const all = await Promise.all(Array.from({ length: 20 }).map(() => login()));

    const res = await login();

    expect(all.filter((x) => x.statusCode === 429)).not.toHaveLength(0);

    expect(res.statusCode).toBe(429);
    expect(res.json()).toMatchSnapshot();
  });

  test('should login', async () => {
    const app = await createServer();

    const res = await app.inject({
      method: 'post',
      url: '/p1/login',
      payload: {
        email: 'treeholechan@gmail.com',
        password: 'lovemeplease',
        'h-captcha-response': 'fake-response',
      },
    });

    expect(res.statusCode).toBe(200);
    // @ts-expect-error remove this ts-ignore after light-my-request release a new version
    expect(res.cookies.filter((x: { name: string }) => x.name === 'sessionID')).toHaveLength(1);
    expect(res.json()).toMatchSnapshot();
  });
});
