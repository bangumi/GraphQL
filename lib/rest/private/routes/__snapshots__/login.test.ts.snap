// Vitest Snapshot v1

exports[`login > should failed on too many requests 1`] = `
{
  "code": "TOO_MANY_REQUESTS",
  "error": "Too Many Requests",
  "message": "too many failed login attempts",
  "statusCode": 429,
}
`;

exports[`login > should login 1`] = `
{
  "id": 382951,
  "nickname": "树洞酱",
  "username": "382951",
}
`;
